### PR TITLE
Add pagination feature to GET charts api

### LINKED ISSUE #270: Is it possible to limit results returned by accessing /api/charts?
We have a lot of charts, would like to have some query strings like ?limit=15&offset=0 for paging, can we add this support? Thanks.

### PR DESCRIPTION
**Pull request by prabhu43**
Fixes #270

### CODE REVIEW

**Comment by prabhu43** [comment]
@jdolitsky Can you review this PR and give your feedback?

File: pkg/chartmuseum/server/multitenant/handlers.go [hunk 1]
```diff
@@ -107,8 +108,20 @@ func (server *MultiTenantServer) getStorageObjectRequestHandler(c *gin.Context)
...
+	}
+	limitString := c.DefaultQuery("limit","20")
+	limit, convErr := strconv.Atoi(limitString)
+	if convErr != nil || limit <= 0 {
+		c.JSON(400, gin.H{"error": "limit is not a valid positive integer"})
+		return
+	}
 	log := server.Logger.ContextLoggingFn(c)
-	allCharts, err := server.getAllCharts(log, repo)
+	allCharts, err := server.getAllCharts(log, repo, offset, limit)
```
Comment (davidovich) [inline pkg/chartmuseum/server/multitenant/handlers.go]: As I understand this, the default values of `offset` and `limit` are respectively `0` and `20`. So if I don't change all my current queries (which do not have these query parameters), I will now get these limits, and this is a backward incompatible change.

File: pkg/chartmuseum/server/multitenant/api.go [hunk 2]
```diff
@@ -19,23 +19,40 @@ package multitenant
...
-func (server *MultiTenantServer) getAllCharts(log cm_logger.LoggingFn, repo string) (map[string]helm_repo.ChartVersions, *HTTPError) {
+func (server *MultiTenantServer) getAllCharts(log cm_logger.LoggingFn, repo string, offset int, limit int) (map[string]helm_repo.ChartVersions, *HTTPError) {
 	indexFile, err := server.getIndexFile(log, repo)
 	if err != nil {
 		return nil, &HTTPError{500, err.Message}
 	}
-	return indexFile.Entries, nil
+	if offset == 0 && limit == -1 {
+		return indexFile.Entries, nil
+	}
```
Comment (davidovich) [inline pkg/chartmuseum/server/multitenant/api.go]: It is only possible to have the default existing behavior by programatically specifying `0, -1` for the offset and limit parameters. This is not possible from the current REST routes.

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 3]
```diff
@@ -837,6 +837,18 @@ func (suite *MultiTenantServerTestSuite) testAllRoutes(repo string, depth int) {
...
+	suite.Equal(200, res.Status(), fmt.Sprintf("200 GET %s/charts", apiPrefix))
+
+	// GET /api/:repo/charts?offset=-1
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts?offset=-1&limit=5", apiPrefix), nil, "")
+	suite.Equal(400, res.Status(), fmt.Sprintf("400 GET %s/charts?offset=-1", apiPrefix))
+
+	// GET /api/:repo/charts?limit=0
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts?offset=-1&limit=5", apiPrefix), nil, "")
+	suite.Equal(400, res.Status(), fmt.Sprintf("400 GET %s/charts?limit=0", apiPrefix))
+
```
Comment (davidovich) [inline pkg/chartmuseum/server/multitenant/server_test.go]: What about adding a test here to assert that a bare `all charts` request returns all candidates (not just the first 20)?

File: pkg/chartmuseum/server/multitenant/handlers.go (hunk 1 above)
Comment (prabhu43) [reply pkg/chartmuseum/server/multitenant/handlers.go]: @davidovich Please check the changes now. I fixed it by setting default limit to "-1"

File: pkg/chartmuseum/server/multitenant/api.go (hunk 2 above)
Comment (prabhu43) [reply pkg/chartmuseum/server/multitenant/api.go]: @davidovich Please check the changes now. Fixed this by setting default limit to "-1" in get all charts api

File: pkg/chartmuseum/server/multitenant/server_test.go (hunk 3 above)
Comment (prabhu43) [reply pkg/chartmuseum/server/multitenant/server_test.go]: @davidovich Can you help me to read response body from response in this test?

**Comment by jdolitsky** [comment]
@prabhu43 thanks for this. I noticed when I ran the tests, coverage was missing for the changes in hanglers.go. It had to do with using the same variable name in a nested scope. I've pushed up a fix, can you please test it on your end?

File: pkg/chartmuseum/server/multitenant/server_test.go (hunk 3 above)
Comment (jdolitsky) [reply pkg/chartmuseum/server/multitenant/server_test.go]: I'm not sure the best way to do this, but doRequest returns a gin.ResponseWriter https://github.com/gin-gonic/gin/blob/master/response_writer.go#L20

**Comment by prabhu43** [comment]
@jdolitsky Thanks for the fix. I tested it and its working. Can you please merge this PR?

**Comment by jdolitsky** [comment]
lgtm

**Pull request merged**